lib/ruby_lsp/requests/type_hierarchy_supertypes.rb

```diff
@@ -0,0 +1,91 @@
+# typed: strict
+# frozen_string_literal: true
+
+module RubyLsp
+  module Requests
+    # ![Type hierarchy supertypes demo](../../type_hierarchy_supertypes.gif)
+    #
+    # The [type hierarchy supertypes
+    # request](https://microsoft.github.io/language-server-protocol/specification#typeHierarchy_supertypes)
+    # displays the list of ancestors (supertypes) for the selected type.
+    #
+    # # Example
+    #
+    # ```ruby
+    # class Foo; end
+    # class Bar < Foo; end
+    #
+    # puts Bar # <-- right click on `Bar` and select "Show Type Hierarchy"
+    # ```
+    class TypeHierarchySupertypes < Request
+      extend T::Sig
+
+      include Support::Common
+
+      sig { params(index: RubyIndexer::Index, item: T::Hash[Symbol, T.untyped]).void }
+      def initialize(index, item)
+        super()
+
+        @index = index
+        @item = item
+      end
+
+      sig { override.returns(T.nilable(T::Array[Interface::TypeHierarchyItem])) }
+      def perform
+        name = @item[:name]
+        entries = @index[name]
+
+        parents = T.let([], T::Array[RubyIndexer::Entry::Namespace])
+        return unless entries&.any?
+
+        entries.each do |entry|
+          next unless entry.is_a?(RubyIndexer::Entry::Namespace)
+
+          if entry.is_a?(RubyIndexer::Entry::Class)
+            parent_class_name = entry.parent_class
+            if parent_class_name
+              resolved_parent_entries = @index.resolve(parent_class_name, entry.nesting)
+              resolved_parent_entries&.each do |entry|
+                next unless entry.is_a?(RubyIndexer::Entry::Class)
+
+                parents << entry
+              end
+            end
+          end
+
+          entry.mixin_operations.each do |mixin_operation|
+            next if mixin_operation.is_a?(RubyIndexer::Entry::Extend)
+
+            mixin_name = mixin_operation.module_name
+            resolved_mixin_entries = @index.resolve(mixin_name, entry.nesting)
+            next unless resolved_mixin_entries
+
+            resolved_mixin_entries.each do |mixin_entry|
+              next unless mixin_entry.is_a?(RubyIndexer::Entry::Module)
+
+              parents << mixin_entry
+            end
+          end
+        end
+
+        parents.uniq.map { |entry| hierarchy_item(entry) }
```

Comment:
With this example:

```rb
class Foo < Object; end
class Foo < Object; end
```

We get two instances of for the `Foo` entries: `Foo#inst1` and `Foo#inst2`.

For each of them we look into the parent:

* Foo#inst1 has `Object` as a superclass name
  * We resolve `Object` and get it's entry instance: `Object#inst3`, we add it to the parents list
* Foo#inst2 has `Object` as a superclass name
  * We resolve `Object` and get it's entry instance: `Object#inst3`, we add it to the parents list again

Now the parents contains `[Object#inst3, Object#inst3]`

In [c9ac6e4](https://github.com/Shopify/ruby-lsp/pull/2103/commits/c9ac6e4158c9d1d59991b368bf2e24680f49e5c0) I actually changed the implementation to use a `Set` rather than having to duplicate.